autokeras/hypermodel/preprocessor.py

```diff
@@ -673,3 +676,233 @@ def set_config(self, config):
+            self.value_counters[cat_col_index1] = collections.defaultdict(
+                return_zero)
+            self.count_frequency[cat_col_index1] = {}
+            for cat_col_index2 in self.categorical_col[index + 1:]:
+                self.categorical_categorical[(
+                    cat_col_index1,
+                    cat_col_index2)] = collections.defaultdict(return_zero)
+            for num_col_index in self.numerical_col:
+                self.numerical_categorical[(
+                    num_col_index,
+                    cat_col_index1)] = collections.defaultdict(return_zero)
+
+    def update(self, x, y=None):
+        self.num_rows += 1
+        x = nest.flatten(x)[0].numpy()
+        for index in range(len(x)):
+            x[index] = x[index].decode('utf-8')
+
+        self._impute(x)
+
+        for col_index in self.categorical_col:
+            key = str(x[col_index])
+            self.label_encoders[col_index].update(key)
+            self.value_counters[col_index][key] += 1
+
+        for col_index1, col_index2 in self.categorical_categorical.keys():
+            key = (str(x[col_index1]), str(x[col_index2]))
+            self.categorical_categorical[(col_index1, col_index2)][key] += 1
+
+        for num_col_index, cat_col_index in self.numerical_categorical.keys():
+            key = str(x[cat_col_index])
+            v = x[num_col_index]
+            self.numerical_categorical[(num_col_index, cat_col_index)][key] += v
+
+    def transform(self, x, fit=False):
+        x = nest.flatten(x)[0].numpy()
+
+        for index in range(len(x)):
+            x[index] = x[index].decode('utf-8')
+        self._impute(x)
+
+        new_values = []
+        # append frequency
+        for col_index in self.high_level1_col:
+            cat_name = str(x[col_index])
+            new_value = self.count_frequency[col_index][cat_name] if \
+                cat_name in self.count_frequency[col_index] else -1
+            new_values.append(new_value)
+
+        # append cat-cat value
+        for key, value in self.high_level_cat_cat.items():
+            col_index1, col_index2 = key
+            pair = (str(x[col_index1]), str(x[col_index2]))
+            new_value = value[pair] if pair in value else -1
+            new_values.append(new_value)
+
+        # append num-cat value
+        for key, value in self.high_level_num_cat.items():
+            num_col_index, cat_col_index = key
+            cat_name = str(x[cat_col_index])
+            new_value = value[cat_name] if cat_name in value else -1
+            new_values.append(new_value)
+
+        # LabelEncoding
+        for col_index in self.categorical_col:
+            key = str(x[col_index])
+            try:
+                x[col_index] = self.label_encoders[col_index].transform(key)
+            except KeyError:
+                x[col_index] = -1
+        return np.hstack((x, np.array(new_values)))
+
+    def _impute(self, x):
```

Comment:
It would be more readable to call it e.g. `fill_missing` since `impute` looks somewhat like `input` and isn't a very common word